cs/src/core/expressions/DeserializerTransform.cs

```diff
@@ -408,6 +424,33 @@ Expression Container(IParser parser, Expression container, Type schemaType, bool
                             }
                         }
 
+                        // For System.Collections.Immutable lists/sets, use the builders to construct them, since ICollection<T>.Add()
+                        // is not supported for them.
+                        var containerGenericTypeDef = container.Type.GetGenericTypeDefinition();
+                        if (immutableListSetTypeNames.Contains(containerGenericTypeDef.FullName))
+                        {
+                            var builderType = container.Type.GetTypeInfo().GetDeclaredNestedType("Builder").MakeGenericType(item.Type);
+                            var builder = Expression.Variable(builderType, container + "_builder");
+                            var builderAdd = builderType.GetMethod(typeof(ICollection<>), "Add", item.Type);
+
+                            addItem = Expression.Block(
+                                Value(valueParser, item, elementType, itemSchemaType, initialize: true),
+                                Expression.Call(builder, builderAdd, item));
+
+                            var toBuilderMethod = container.Type.FindMethod("ToBuilder");
+                            var toImmutableMethod = builderType.FindMethod("ToImmutable");
+                            var constructBuilder = Expression.Assign(builder, Expression.Call(container, toBuilderMethod));
+                            var reconstructImmutable = Expression.Assign(container, Expression.Call(builder, toImmutableMethod));
+
+                            return Expression.Block(
+                                new[] { item, builder },
+                                constructBuilder,
+                                beforeLoop,
+                                ControlExpression.While(next, addItem),
+                                afterLoop,
+                                reconstructImmutable);
+                        }
+
```

Comment:
Minor: could you put the `ICollection.Add` code path in an `else` branch of `if (immutableListSetTypeNames.Contains(containerGenericTypeDef.FullName))`?

Reply:
Done, also conformed to the pattern of just defining the necessary variables so there's a single return at the end